File: leetcode_75/Linked_List/python/2095.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def __get_length(self, head: Optional[ListNode]) -> int:
        cur = head
        n = 1
        while cur.next is not None:
            cur = cur.next
            n += 1

        return n

    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        n = self.__get_length(head)
        if n == 1:
            return

        middle = n // 2

        prev = None
        cur = head
        next = cur.next
        k = 0
        while k != middle:
            prev = cur
            cur = next
            next = cur.next
            k += 1


        if next is not None:
            prev.next = next
        else:
            prev.next = None

        return head

File: leetcode_75/Linked_List/cpp/2095.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *deleteMiddle(ListNode *head)
    {
        int n = this->get_length(head);
        if (n == 1)
        {
            return nullptr;
        }
        int middle = n / 2;

        ListNode *prev = nullptr;
        ListNode *cur = head;
        ListNode *next = cur->next;
        int k = 0;
        while (k != middle)
        {
            prev = cur;
            cur = next;
            next = cur->next;
            k++;
        }

        if (next != nullptr)
        {
            prev->next = next;
        }
        else
        {
            prev->next = nullptr;
        }

        return head;
    }

private:
    int get_length(ListNode *head)
    {
        ListNode *cur = head;
        int n = 1;
        while (cur->next != nullptr)
        {
            cur = cur->next;
            n++;
        }
        return n;
    }
};